### PR TITLE
Integration of Sprint 6 and Sprint 7 Changes with Optimistic Updates and UI Enhancements

### PR DESCRIPTION
**Pull request by TuanDangNguyen2003**
### PR Description: Sprint 6 and Sprint 7 Changes

This PR combines the changes from **Sprint 6** and **Sprint 7**. 

Due to major changes in the module, the PR for Sprint 6 was not merged into the main branch. Instead, the changes were carried forward to Sprint 7, where they were fixed, adapted to include **optimistic updates**, and thoroughly tested. The tasks accomplished for each sprint are detailed below:

#### **Sprint 6 Tasks**:
1. Redesigned the modal for each card to ensure consistent popup behavior.
2. Added a loading state to all buttons to prevent spamming during action execution.
3. Improved the UI for deleting or editing, making the process more approachable.
4. Implemented testing for new features, including modal functionality.
5. Fixed and added mocks for a feature involving optimistic updates.

#### **Sprint 7 Tasks**:
1. Integrated optimistic updates across the entire module.
2. Refactored tests to ensure compatibility with new features.
3. Added card playing options for users.
4. Enhanced the UI by converting the playing status display into an icon for a better visual experience.

All tasks have been thoroughly tested, and the module is now stable and ready for merging into the main branch.

### CODE REVIEW

File: edweiss-app/app/(app)/deck/[id]/card/creation/index.tsx [hunk 1]
```diff
@@ -37,95 +38,111 @@ import { DecksRepository } from '../../../_layout';
...
-		router.back();
-	}
-
-	return (
-		<>
-			<RouteHeader title='Create a card' />
-
-			<TScrollView>
-				{/* <TextInput value={deckName} onChangeText={n => setDeckName(n)} placeholder='Deck name' placeholderTextColor={'#555'} style={{ backgroundColor: Colors.dark.crust, borderColor: Colors.dark.blue, borderWidth: 1, padding: 8, paddingHorizontal: 16, margin: 16, marginBottom: 0, color: 'white', borderRadius: 14, fontFamily: "Inter" }}>
+    const { deckId } = useStringParameters();
```
Comment (FerdinandGomerChiu) [inline edweiss-app/app/(app)/deck/[id]/card/creation/index.tsx:41]: Out of curiosity, I am seeing a huge chunk of change (red color), but the new code seems mostly similar (except for checks). Do you know why is that?

Comment (TuanDangNguyen2003) [reply edweiss-app/app/(app)/deck/[id]/card/creation/index.tsx:41]: There is a commented code in { /* ... */} so I just deleted it I think.

Comment (TuanDangNguyen2003) [reply edweiss-app/app/(app)/deck/[id]/card/creation/index.tsx:41]: Ah so first I added some "useStates", then I adapted optimistic update feature. There are many changes so the editor probably assume that I just deleted everything before and make something totally new.

Comment (TuanDangNguyen2003) [reply edweiss-app/app/(app)/deck/[id]/card/creation/index.tsx:41]: Ah I see, one of the thing is the indentation is not the same.

File: edweiss-app/components/memento/CardScreenComponent.tsx [hunk 2]
```diff
@@ -86,69 +115,38 @@ const CardScreenComponent: ReactComponent<{ deckId: string, cardIndex: number; i
...
     // Delete card
-    async function deleteCard() {
-
-        const res = await callFunction(Memento.Functions.deleteCards, { deckId: deckId, cardIndices: [cardIndex] });
-        if (res.status == 1) {
-            console.log(`Card deleted with id ${res.data.id}`);
-            isModal ? modalRef.current?.dismiss() : router.back();
-        }
-
+    const deleteCard = async () => {
```
Comment (FerdinandGomerChiu) [inline edweiss-app/components/memento/CardScreenComponent.tsx:119]: Why did you transform a function into this lambda?

File: edweiss-app/components/memento/DeleteDeckModalAction.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import ReactComponent from '@/constants/Component';
+import { BottomSheetModalMethods } from '@gorhom/bottom-sheet/lib/typescript/types';
+import { useState } from 'react';
+import TView from '../core/containers/TView';
+import ModalContainer from '../core/modal/ModalContainer';
+import TText from '../core/TText';
+import FancyButton from '../input/FancyButton';
+
+export const DeleteOptionModalDisplay: ReactComponent<{
```
Comment (FerdinandGomerChiu) [inline edweiss-app/components/memento/DeleteDeckModalAction.tsx:9]: Great job on separating your components in different files: this is more readable than what I did (putting every modal component in the same file).

File: edweiss-app/utils/memento/utilsFunctions.tsx [hunk 4]
```diff
@@ -18,31 +39,65 @@ export const getStatusColor = (status: string) => {
...
+ * 
+ * @param {boolean} isDupplicate - Check if the question is duplicated
+ * @param {boolean} isEmptyField - Check if the question field is empty
+ * @param {React.Dispatch<React.SetStateAction<boolean>>} setExistedQuestion - Set the state of the question existence
+ * @param {React.Dispatch<React.SetStateAction<boolean>>} setEmptyField - Set the state of the empty field
+ * @param {React.Dispatch<React.SetStateAction<boolean>>} setIsLoading - Set the state of the loading status
+ * 
+ * @returns {number} - 0 if the question is duplicated or the field is empty, 1 otherwise
+ */
+export const checkDupplication_EmptyField = (
```
Comment (FerdinandGomerChiu) [inline edweiss-app/utils/memento/utilsFunctions.tsx:76]: Why return 0 or 1 instead of a boolean? Do you plan on returning other values in the future?

File: edweiss-app/components/memento/CardScreenComponent.tsx (hunk 2 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/components/memento/CardScreenComponent.tsx:119]: So that I can pass this function to a component that requires a function as its mandatory argument.

File: edweiss-app/components/memento/DeleteDeckModalAction.tsx (hunk 3 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/components/memento/DeleteDeckModalAction.tsx:9]: xDD thank you. It took me some times to organize things but afterwards, it is more readable.

File: edweiss-app/__test__/memento/card.test.tsx [hunk 5]
```diff
@@ -227,41 +235,88 @@ describe('CardListScreen', () => {
...
+			expect(mockGoToPath).not.toHaveBeenCalled(); // Ensure goToPath is not called
 	});
 
 	// Add additional tests for edge cases if necessary
 });
+			*/
 
+/*
 describe('handleCardPress', () => {
+				let modalRef: React.RefObject<BottomSheetModal>;
```
Comment (FerdinandGomerChiu) [inline edweiss-app/__test__/memento/card.test.tsx:274]: I think pushing commented code is dangerous: it might take away some points in your grading (I was victim of this during MS1) . I also told @holoisme that. We might have to check if this is the case with the coaches though.

File: edweiss-app/utils/memento/utilsFunctions.tsx (hunk 4 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/utils/memento/utilsFunctions.tsx:76]: Hm ... I forgot that we can return a boolean. But I guess in the future, if there is some edge cases, there might be exception so it will return different values for each cases.

File: edweiss-app/components/memento/CardScreenComponent.tsx (hunk 2 above)
Comment (FerdinandGomerChiu) [reply edweiss-app/components/memento/CardScreenComponent.tsx:119]: Hmm I believe you can still do this with a regular function. (In fact, I use this a lot)

File: edweiss-app/__test__/memento/card.test.tsx (hunk 5 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/__test__/memento/card.test.tsx:274]: I will leave it out in a .txt file I guess. Because I want to have an example of use case when testing.

File: edweiss-app/utils/memento/utilsFunctions.tsx (hunk 4 above)
Comment (FerdinandGomerChiu) [reply edweiss-app/utils/memento/utilsFunctions.tsx:76]: Still hung up on C, I see.. (jk lol)

File: edweiss-app/components/memento/CardScreenComponent.tsx (hunk 2 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/components/memento/CardScreenComponent.tsx:119]: You are right but since this implementation still works well so I will keep in mind your suggestion for future modification.

**Comment by FerdinandGomerChiu** [comment]
I'll add a single comment for the testing, as it is quite systematic:
Great job on getting familiar with @holoisme 's Repository. This is quite the feat, and will definitely prove to be useful for everyone.

**Comment by TuanDangNguyen2003** [comment]
Thank you :DD

File: edweiss-app/utils/memento/utilsFunctions.tsx (hunk 4 above)
Comment (FerdinandGomerChiu) [reply edweiss-app/utils/memento/utilsFunctions.tsx:76]: By the way, if you need to return more than 2 types, consider using a custom type, i.e returnType = "valid"|"error_one"|"error_two". The intention is clearer than when you use a number.

**Comment by FerdinandGomerChiu** [comment]
I see you are using lots of modals. My last PR added scrollable modals, and adjusted colors if you want to take a look.

**Review by FerdinandGomerChiu — APPROVED** [review]
Overall, this is a lot of work. Good job on coverage and code quality. Keep it up! (Don't forget the wiki, if you haven't already put your mocks in it)

File: edweiss-app/utils/memento/utilsFunctions.tsx (hunk 4 above)
Comment (TuanDangNguyen2003) [reply edweiss-app/utils/memento/utilsFunctions.tsx:76]: Oh nice idea. Thank you :D

**Comment by TuanDangNguyen2003** [comment]
Thank you. I will put the mock in the Wiki.
